### PR TITLE
8292375: Convert ProtectionDomainCacheTable to ResourceHashtable

### PR DESCRIPTION
**Pull request by coleenp**
Please review this simple conversion for the ProtectionDomainCacheTable from Old Hashtable to ResourceHashtable.  There are specific tests for this table in test/hotspot/jtreg/runtime/Dictionary and serviceability/dcmd/vm/DictionaryStatsTest.java.
Also tested with tier1-7.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8292375](https://bugs.openjdk.org/browse/JDK-8292375): Convert ProtectionDomainCacheTable to ResourceHashtable

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Ioi Lam](https://openjdk.org/census#iklam) (@iklam - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/10043/head:pull/10043` \
`$ git checkout pull/10043`

Update a local copy of the PR: \
`$ git checkout pull/10043` \
`$ git pull https://git.openjdk.org/jdk pull/10043/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 10043`

View PR using the GUI difftool: \
`$ git pr show -t 10043`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/10043.diff">https://git.openjdk.org/jdk/pull/10043.diff</a>

</details>

### CODE REVIEW

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 1]
```diff
@@ -35,23 +35,24 @@
...
 #include "runtime/atomic.hpp"
+#include "runtime/mutexLocker.hpp"
 #include "utilities/growableArray.hpp"
-#include "utilities/hashtable.inline.hpp"
+#include "utilities/resourceHash.hpp"
 
-unsigned int ProtectionDomainCacheTable::compute_hash(Handle protection_domain) {
-  // Identity hash can safepoint, so keep protection domain in a Handle.
-  return (unsigned int)(protection_domain->identity_hash());
+unsigned int ProtectionDomainCacheTable::compute_hash(const WeakHandle& protection_domain) {
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:42]: We are we now using `WeakHandle` everywhere?

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 2]
```diff
@@ -35,23 +35,24 @@
...
+  return (unsigned int)(protection_domain.resolve()->identity_hash());
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:43]: And if it is a `WeakHandle` can't `resolve` now return NULL?

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 3]
```diff
@@ -128,108 +159,69 @@ void ProtectionDomainCacheTable::unlink() {
...
-      }
-      entry = *p;
-    }
-  }
   _total_oops_removed += oops_removed;
   _dead_entries = false;
 }
 
-void ProtectionDomainCacheTable::print_on(outputStream* st) const {
+void ProtectionDomainCacheTable::print_on(outputStream* st) {
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:168]: It is a little disconcerting that `print_on` can no longer be a `const` function!

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 4]
```diff
@@ -35,23 +35,24 @@
...
+bool ProtectionDomainCacheTable::equals(const WeakHandle& protection_domain1, const WeakHandle& protection_domain2) {
+  return protection_domain1.peek() == protection_domain2.peek();
 }
 
-ProtectionDomainCacheTable::ProtectionDomainCacheTable(int table_size)
-  : Hashtable<WeakHandle, mtClass>(table_size, sizeof(ProtectionDomainCacheEntry))
-{   _dead_entries = false;
-    _total_oops_removed = 0;
-}
+ResourceHashtable<WeakHandle, WeakHandle, 1009, ResourceObj::C_HEAP, mtClass,
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:50]: I am struggling to understand what the key and values types are in this hashtable ???

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 5]
```diff
@@ -128,108 +159,69 @@ void ProtectionDomainCacheTable::unlink() {
 
   // Purge any deleted entries outside of the SystemDictionary_lock.
   purge_deleted_entries();
+  int oops_removed = purge_entries_from_table();
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:162]: Maybe add comment

`int oops_removed = purge_entries_from_table(); // reacquires SD lock`

otherwise it gives the false impression this is done lock-free.

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 6]
```diff
@@ -128,108 +159,69 @@ void ProtectionDomainCacheTable::unlink() {
...
-oop ProtectionDomainCacheEntry::object() {
-  return literal().resolve();
+  auto verifier = [&] (WeakHandle& key, WeakHandle& value) {
+    guarantee(value.peek() == nullptr || oopDesc::is_oop(value.peek()), "must be an oop");
+  };
+  _pd_cache_table.iterate_all(verifier);
 }
 
 // The object_no_keepalive() call peeks at the phantomly reachable oop without
 // keeping it alive. This is okay to do in the VM thread state if it is not
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:186]: You don't call `object_no_keepalive()` any more

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 7]
```diff
@@ -128,108 +159,69 @@ void ProtectionDomainCacheTable::unlink() {
...
-      return e;
+  WeakHandle w(Universe::vm_weak(), protection_domain);
+  bool created;
+  WeakHandle* wk = _pd_cache_table.put_if_absent(w, w, &created);
+  if (!created) {
+    // delete the one created since we already had it in the table
+    w.release(Universe::vm_weak());
+    // Keep entry alive
+    (void)wk->resolve();
+    return *wk;
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:202]: Can't this be factored out of the if-else as `wk` is always the right entry to resolve and return?

**Review by dholmes-ora — COMMENTED** [review]
Hi Coleen,

Probably due to my lack of understanding of the existing code I found this conversion very hard to follow. A number of comments below.

Thanks.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:42]: WeakHandle is the object we're storing as in the hashtable.  It also turns out to be the key.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:43]: compute_hash() is always called on a live WeakHandle so will never return null.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 3 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:168]: It's static, so it can't.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:50]: WeakHandle is the key and the value in this hashtable.  We need to match the WeakHandle in the input by value (see equals function), but we also want to return the WeakHandle to store in to the pd_set list in the DictionaryEntry.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:162]: Ok.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 6 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:186]: This one (not the one I removed), is called by dictionary.cpp - the pd_set is a linked list of ProtectionDomainEntry, where we don't keep the WeakHandle alive when looking at the value.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 7 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:202]: Yes, that works.

**Comment by coleenp** [comment]
Thanks for reading through this.  I was going to make a larger change, then changed my mind.  This conversion is limited.  It is able to take advantage to the ability to copy WeakHandle without side effects, so that we don't have to store ProtectionDomainCacheEntry into the pd_set linked list, which makes it nicer.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 2 above)
Comment (iklam) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:43]: I think you should add an assert that the `protection_domain.resolve()` never returns null, with a comment explaining why.

File: src/hotspot/share/classfile/protectionDomainCache.hpp [hunk 8]
```diff
@@ -28,88 +28,42 @@
...
-      Hashtable<WeakHandle, mtClass>::new_entry(hash, protection_domain);
-    return entry;
-  }
-
-  static unsigned int compute_hash(Handle protection_domain);
-
-  int index_for(Handle protection_domain);
-  ProtectionDomainCacheEntry* add_entry(int index, unsigned int hash, Handle protection_domain);
-  ProtectionDomainCacheEntry* find_entry(int index, Handle protection_domain);
+class ProtectionDomainCacheTable : public AllStatic {
```
Comment (iklam) [inline src/hotspot/share/classfile/protectionDomainCache.hpp:35]: How about adding a comment to say what this table maps from/to? Something like:
```
// The ProtectionDomainCacheTable maps all java.security.ProtectionDomain objects that are
// registered by DictionaryEntry::add_protection_domain() to a unique WeakHandle. 
```

**Review by iklam — APPROVED** [review]
Looks good overall and seems to be equivalent to the old code. Just a couple of nits.

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 8 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:35]: Okay, I added your comment.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:43]: I agree. In general if code is using WeakHandle's then I would expect to see null checks.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 3 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:166]: Ah I see - I hadn't noticed it was now static.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 6 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:184]: I don't understand your point. My point is that the comment no longer seems correct.

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 8 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:36]: I still can't decipher what this "map" actually holds. What does the unique WeakHandle actually refer to?

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 6 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:184]: The first sentence of the comment is correct.  I don't know what the second sentence means, so I'll remove it.

File: src/hotspot/share/classfile/protectionDomainCache.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.cpp:45]: I added the comment.

File: src/hotspot/share/classfile/protectionDomainCache.hpp [hunk 9]
```diff
@@ -29,7 +29,8 @@
 #include "oops/weakHandle.hpp"
 #include "runtime/atomic.hpp"
 
-// The ProtectionDomainCacheTable contains all protection domain oops as WeakHandles.
+// The ProtectionDomainCacheTable maps all java.security.ProtectionDomain objects that are
+// registered by DictionaryEntry::add_protection_domain() to a unique WeakHandle.
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.hpp:33]: Now that I understand what this table does, this comment is confusing to me. The table maps each PD to itself (using the WeakHandle as the actual key and value) as a means to track which PDs have been seen/registered. I would describe it thus:
```
// The ProtectionDomainCacheTable records all of the java.security.ProtectionDomain instances that have
// been registered by DictionaryEntry::add_protection_domain(). We use a hashtable to
map each  PD 
// instance to itself (using WeakHandles) to allow for fast lookup.
```

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 8 above)
Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:36]: I've suggested a different comment above.

File: src/hotspot/share/classfile/protectionDomainCache.cpp [hunk 10]
```diff
@@ -129,107 +134,91 @@ void ProtectionDomainCacheTable::unlink() {
...
 }
 
 // The object_no_keepalive() call peeks at the phantomly reachable oop without
-// keeping it alive. This is okay to do in the VM thread state if it is not
-// leaked out to become strongly reachable.
-oop ProtectionDomainCacheEntry::object_no_keepalive() {
-  return literal().peek();
-}
-
+// keeping it alive.  This is used for traversing DictionaryEntry pd_set.
```
Comment (dholmes-ora) [inline src/hotspot/share/classfile/protectionDomainCache.cpp:186]: My confusion here was that the comment belonged to the `ProtectionDomainCacheEntry::object_no_keepalive` method, but you have deleted that and are now applying the comment to the `ProtectionDomainEntry::object_no_keepalive()` method.
The end result is okay.

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 9 above)
Comment (iklam) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: “Mapping to itself” is how this table works internally. I am not sure if this information is useful here. But the purpose of this table is really to keep track of all PDs that have been registered and not yet garbage collected.

Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: "mapping to itself" is more useful than "mapping ... to a unique Weakhandle" - which is even more of an internal implementation detail. I found the use of this table very hard to discern based on the internal use of the hashtable, as there is no real mapping operation - we simply track if a PD has been seen or not. The use of the hashtable is purely for lookup convenience - we could instead have a linked-list of PD's that we traverse for lookup.
So perhaps we drop my second sentence above, and move it to where the hashtable itself is declared?

**Comment by iklam** [comment]
I am utterly confused about why we need ProtectionDomainCacheTable at all. The only interface between this class and the rest of the world is:

```
DictionaryEntry::add_protection_domain() {
 
    WeakHandle obj = ProtectionDomainCacheTable::add_if_absent(protection_domain);
    // Additions and deletions hold the SystemDictionary_lock, readers are lock-free
    ProtectionDomainEntry* new_head = new ProtectionDomainEntry(obj, _pd_set);
}
```

(and there's code elsewhere for cleaning up this table, but that wouldn't be necessary if no one calls `add_if_absent`!).

Why doesn't DictionaryEntry::add_protection_domain allocate the WeakHandle itself?

I am looking at the JDK 8 code. It seems like ProtectionDomainCacheTable was needed before we had WeakHandle, so we had to do all the reference management by hand:

https://hg.openjdk.java.net/jdk8/jdk8/hotspot/file/87ee5ee27509/src/share/vm/classfile/dictionary.hpp#l137

But for today, is ProtectionDomainCacheTable a relic that can be thrown away?

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 9 above)
Comment (iklam) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: "mapping a PD to a unique Weakhandle" is not an implementation detail. It's the only useful API provided by this class:

```
WeakHandle obj = ProtectionDomainCacheTable::add_if_absent(protection_domain);
```

and that's the reason I question why this table is needed at all.

Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: The fact it is a weakhandle is an implementation detail. The table simply records whether a PD (wrapped in a WeakHandle) has been seen.

**Comment by dholmes-ora** [comment]
> I am utterly confused about why we need ProtectionDomainCacheTable at all.

It tracks whether we have seen this PD before so that `validate_protection_domain` can skip the Java upcall to `checkPackageAccess`.

**Comment by coleenp** [comment]
The reason we have the table is that you can have the same instance of java.security.ProtectionDomain used on multiple DictionaryEntry._pd_set lists.  That is, multiple classes in the class loader data have had their loading initiated with the same protection domain, which is common.  Creating WeakHandles for each is redundant and wasteful, so we want to have one WeakHandle for multiple entries to point to.  The table is a place to store it.  I had another version that stored the WeakHandle entries in the ClassLoaderData of the Dictionary of loaded classes that had them in their pd_set, but they had a linear lookup to determine if they were unique. We even performance tested that.  In the end, I just translated the table to ResourceHashtable to make it easier to review and understand what this does.

As an example, if you run hello world with the security manager allowed, you see this:

java -Xlog:protectiondomain=trace -Djava.security.manager=allow -cp ~/work Hello
[0.188s][trace][protectiondomain] adding protection domain for class java/lang/Object class loader: a 'jdk/internal/loader/ClassLoaders$AppClassLoader'{0x00000007fef58f90} protection domain: a 'java/security/ProtectionDomain'{0x000000011f01d820} pd set count = #1
[0.189s][trace][protectiondomain] adding protection domain for class java/lang/String class loader: a 'jdk/internal/loader/ClassLoaders$AppClassLoader'{0x00000007fef58f90} protection domain: a 'java/security/ProtectionDomain'{0x000000011f01d820} pd set count = #1
[0.189s][trace][protectiondomain] adding protection domain for class java/lang/Exception class loader: a 'jdk/internal/loader/ClassLoaders$AppClassLoader'{0x00000007fef58f90} protection domain: a 'java/security/ProtectionDomain'{0x000000011f01d820} pd set count = #1
[0.189s][trace][protectiondomain] adding protection domain for class java/lang/System class loader: a 'jdk/internal/loader/ClassLoaders$AppClassLoader'{0x00000007fef58f90} protection domain: a 'java/security/ProtectionDomain'{0x000000011f01d820} pd set count = #1
[0.189s][trace][protectiondomain] adding protection domain for class java/io/PrintStream class loader: a 'jdk/internal/loader/ClassLoaders$AppClassLoader'{0x00000007fef58f90} protection domain: a 'java/security/ProtectionDomain'{0x000000011f01d820} pd set count = #1
Hello5

Whenever we initiate loading (for example) java.lang.Object in a non bootstrap class loader with the SecureClassLoader, we pass this protection domain.  Notice how the same protection domain is added to each class.

In JDK8 we pointed to the oop itself which required the GC to walk the global SystemDictionary.  We definitely never want that, so moved it to a table, and made several other changes to enable concurrent class unloading after that - moved the Dictionaries of loaded classes to each class loader data.

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 9 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: Since I know what this table does, either comment is fine if it helps someone understand it.
The "map each PD to itself" is pretty odd to me too.   How about collect each PD for fast lookup in a hashtable?  The code says how it's mapped.  There isn't that much code and it's easy to see how the Key is mapped.

Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: // The ProtectionDomainCacheTable maps all java.security.ProtectionDomain objects that are
// registered by DictionaryEntry::add_protection_domain() to a unique entry.  The entry
// is a WeakHandle that holds the protection domain oop.

or points to....  either is accurate.

Comment (dholmes-ora) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: It is the "maps ... to a unique entry" that I find most problematic - it begs the question as to what the unique entry is, when in reality it maps a PD instance (wrapped in a WeakHandle) to itself (wrapped in a WeakHandle).

For the sake of progress, approved.

**Review by dholmes-ora — APPROVED** [review]

**Review by iklam — APPROVED** [review]

File: src/hotspot/share/classfile/protectionDomainCache.hpp (hunk 9 above)
Comment (coleenp) [reply src/hotspot/share/classfile/protectionDomainCache.hpp:33]: The "unique" means there's only one value in the table for this protection domain oop.  So the oop is unique.  The use of word "mapping" might be what's confusing.
If we ever have to visit this code again (which I hope not, except to remove it because the security manager is finally removed since not many use it), we can see if this comment still makes sense to at least some of us.
Thanks for approving for progress.
